### PR TITLE
Add support for MS Visual Studio

### PR DESCRIPTION
**Pull request by JeremyLWright**
Added support for Visual Studio.
Added travis-ci build configuration.
Added cmake build script for building with GCC, Clang, and Visual Studio
Integrated fixes from another fork.
Added reporter for catch testing library in addition to gtest.

### CODE REVIEW

File: include/autocheck/generator_combinators.hpp [hunk 1]
```diff
@@ -45,10 +46,11 @@ namespace autocheck {
           pred(pred), gen(gen) {}
 
         result_type operator() (size_t size = 0) {
-          while (true) {
+          for(size_t i{0}; i < std::numeric_limits<size_t>::max(); ++i){
```
Comment (thejohnfreeman) [inline include/autocheck/generator_combinators.hpp]: What is the logic here exactly? I don't think the filter promises to halt if the predicate accepts nothing, and in general, no generator should call a sub-generator (including itself) with a larger size than it was given.

File: include/autocheck/value.hpp [hunk 2]
```diff
@@ -17,10 +17,14 @@ namespace autocheck {
         Heap
       }    allocation = None;
 
+#ifndef _MSC_VER
```
Comment (thejohnfreeman) [inline include/autocheck/value.hpp]: Please include a comment why this is necessary.

File: include/autocheck/reporter.hpp [hunk 3]
```diff
@@ -92,8 +92,76 @@ namespace autocheck {
       }
   };
 
+  class msg_reporter : public reporter {
```
Comment (thejohnfreeman) [inline include/autocheck/reporter.hpp]: What's going on here?

I think msg_reporter can be removed and gtest_checker can use ostream_reporter with ostringstream.

File: include/autocheck/generator_combinators.hpp (hunk 1 above)
Comment (JeremyLWright) [inline include/autocheck/generator_combinators.hpp]: I merged this from another fork and didn't fully understand the generator at the time.

Comment (thejohnfreeman) [inline include/autocheck/generator_combinators.hpp]: I see. If it makes no difference to you, I'd rather revert this change.

File: include/autocheck/value.hpp (hunk 2 above)
Comment (JeremyLWright) [inline include/autocheck/value.hpp]: Done.

**Comment by JeremyLWright** [comment]
I incorporated your comments.
Once you accept this merge request, I think you'll need to create a travis-ci build under your account and change the README file to point to it.

File: include/autocheck/reporter.hpp [hunk 4]
```diff
@@ -92,8 +92,48 @@ namespace autocheck {
...
+struct gtest_checker
+{
+    std::function<void(std::string const&)> expectfn;
+    std::string msg;
+    ostream_reporter rep;
+    template<class FN>
+    void operator <<(FN const& fn)
+    {
+        fn(rep);
+        expectfn(msg);
```
Comment (thejohnfreeman) [inline include/autocheck/reporter.hpp]: This needs to read the string from the ostringstream wrapped by rep.

File: test/generator.cpp [hunk 5]
```diff
@@ -18,7 +18,7 @@ void generate() {
     using namespace autocheck;
     std::clog << gen(i);
   }
-  std::clog << std::endl;
+  std::clog << '\n';
```
Comment (thejohnfreeman) [inline test/generator.cpp]: Why this change?

File: tutorial/factorial/factorial.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#define CATCH_CONFIG_MAIN
+#include <catch.hpp>
+
+#include <autocheck/sequence.hpp>
+#include <autocheck/check.hpp>
+
+std::size_t Factorial(const std::size_t number)
+{
```
Comment (thejohnfreeman) [inline tutorial/factorial/factorial.cpp]: Opening braces on preceding lines all throughout this file, please.

File: test/catch_test.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  SECTION("bool with default parameters."){
+    ac::check<bool>([] (bool x) { return true; });
+  }
+
+  reverse_prop_t reverse_prop;
+  SECTION("reverse property over int"){
+
+  ac::check<std::vector<int>>(reverse_prop, 100,
```
Comment (thejohnfreeman) [inline test/catch_test.cpp]: I believe the lines in this `SECTION` and the next two need to be indented another level.

File: include/autocheck/reporter.hpp [hunk 8]
```diff
@@ -92,8 +92,48 @@ namespace autocheck {
       }
   };
 
+struct gtest_checker
+{
```
Comment (thejohnfreeman) [inline include/autocheck/reporter.hpp]: Braces on preceding lines please.

File: test/catch_test.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      b = xs.rbegin(), a = b++, e = xs.rend(); (b != e) && (*a < *b); ++b, ++a)
+  {
+    std::iter_swap(a, b);
+  }
+}
+
+TEST_CASE("Archetype Testing", "[check]") {
+  ac::catch_reporter rep;
+
+  SECTION("bool"){
```
Comment (thejohnfreeman) [inline test/catch_test.cpp]: Space between `)` and `{` please, throughout this file.

File: test/generator.cpp (hunk 5 above)
Comment (JeremyLWright) [inline test/generator.cpp]: http://llvm.org/docs/CodingStandards.html#avoid-std-endl

Comment (thejohnfreeman) [inline test/generator.cpp]: I was under the impression `'\n'` was not portable, but it looks like it is for [text mode](http://stackoverflow.com/questions/1911018/what-does-stdendl-represent-exactly-on-each-platform).

File: include/autocheck/reporter.hpp (hunk 4 above)
Comment (thejohnfreeman) [inline include/autocheck/reporter.hpp]: I was going to write out what I think this code should be, but it appears it isn't used? If not, please remove it.

**Comment by thejohnfreeman** [comment]
Almost there. I didn't get the emails for the latest fixes again... Is there some option I need to tick to get notified of new commits?

**Comment by JeremyLWright** [comment]
Hello John, 
Let me know if there is anything else in this pull request. I'm going to start writing more tutorials. Are there any good references you have on the subject of randomized testing? Or anything that inspired this design of autocheck?

Jeremy

**Comment by thejohnfreeman** [comment]
It's been a while since I did this research, but these resources look familiar:

Original source:
http://www.cse.chalmers.se/~rjmh/QuickCheck/manual.html
Very short introduction. Even more links:
https://wiki.haskell.org/Introduction_to_QuickCheck1
Good looking introduction. Haven't read it, but enjoyed other things I've read on their site:
https://www.fpcomplete.com/user/pbv/an-introduction-to-quickcheck-testing
SmallCheck. Links in README:
https://hackage.haskell.org/package/smallcheck

For implementation, I was definitely inspired by the other C++ QuickCheckers linked in my [acknowledgments](https://github.com/thejohnfreeman/autocheck/wiki#acknowledgments), though I think I deviated somewhat to make it feel closer to Haskell.

**Comment by thejohnfreeman** [comment]
Thank you for the contribution!

**Comment by JeremyLWright** [comment]
Cool thank you for accepting!
